Updating and publishing databases without `parquet` fails with 1.7.2

I am using Python 3.10, `audb==1.7.2`.

```bash
Traceback (most recent call last):
  File "/home/user/mydb/2.0.0/publish.py", line 15, in <module>
    audb.publish(
  File "/home/user/mydb/2.0.0/myenv/lib/python3.10/site-packages/audb/core/publish.py", line 698, in publish
    raise RuntimeError(
RuntimeError: You want to depend on '1.0.0' of mydb, but the dependency file 'db.parquet' in /home/user/mydb/build does not match the dependency file for the requested version in the repository. Did you forgot to call 'audb.load_to(/home/user/mydb/build, mydb, version='1.0.0') or modified the file manually?
```

Switching back to `audb==1.6.5`, where no `db.parquet` file is generated for the updated database made it work.

The corresponding database has only `misc_table` (in case this is relevant).

The problem arises from our implementation of `audb.Dependencies.__eq__()`, which compares the dataframes of the dependency tables:

https://github.com/audeering/audb/blob/44df511ff8709b505e4cc3cad44ac74349bb8567/audb/core/dependencies.py#L107-L117

When loading the original dependency table of version `1.4.1`, that was stored with `audb<1.7` its string dtype is `string[python]`, whereas the new dependency table has `string[pyarrow]`, which leads to the following results when asserting the dataframes should be equal:

```
AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="archive") are different

Attribute "dtype" are different
[left]:  string[python]
[right]: string[pyarrow]
```

So, I guess we should update the implementation of `audb.Dependencies.__eq__()` to ensure backward compatibility.

--- a/audb/core/dependencies.py
+++ b/audb/core/dependencies.py
@@ -108,7 +108,7 @@ class Dependencies:
             ``True`` if both dependency tables have the same entries
 
         """
-        return self._df.equals(other._df.astype(self._df.dtypes))
+        return self._df.equals(other._df)
 
     def __getitem__(self, file: str) -> typing.List:
         r"""File information.

--- a/audb/core/dependencies.py
+++ b/audb/core/dependencies.py
@@ -586,9 +586,10 @@ class Dependencies:
             with correct dtypes
 
         """
-        # Check if we have an old cache,
-        # by checking the dtype of index,
-        # which changed to `object` in version 1.7.0
+        # Check the dtype of index,
+        # to decide if we need to update dtypes,
+        # as dtype of index changed to `object`
+        # in version 1.7.0 of audb.
         if df.index.dtype != define.DEPEND_INDEX_DTYPE:
             df.index = df.index.astype(define.DEPEND_INDEX_DTYPE, copy=False)
             columns = define.DEPEND_FIELD_NAMES.values()

--- a/audb/core/dependencies.py
+++ b/audb/core/dependencies.py
@@ -57,14 +57,8 @@ class Dependencies:
     """  # noqa: E501
 
     def __init__(self):
-        data = {}
-        for name, dtype in zip(
-            define.DEPEND_FIELD_NAMES.values(),
-            define.DEPEND_FIELD_DTYPES.values(),
-        ):
-            data[name] = pd.Series(dtype=dtype)
-        self._df = pd.DataFrame(data)
-        self._df.index = self._df.index.astype(define.DEPEND_INDEX_DTYPE)
+        self._df = pd.DataFrame(columns=define.DEPEND_FIELD_NAMES.values())
+        self._df = self._set_dtypes(self._df)
         # pyarrow schema
         # used for reading and writing files
         self._schema = pa.schema(
@@ -114,7 +108,7 @@ class Dependencies:
             ``True`` if both dependency tables have the same entries
 
         """
-        return self._df.equals(other._df)
+        return self._df.equals(other._df.astype(self._df.dtypes))
 
     def __getitem__(self, file: str) -> typing.List:
         r"""File information.
@@ -328,12 +322,10 @@ class Dependencies:
             )
         if extension == "pkl":
             self._df = pd.read_pickle(path)
-            # Correct dtype of index
+            # Correct dtypes
             # to make backward compatiple
             # with old pickle files in cache
-            # that might use `string` as dtype
-            if self._df.index.dtype != define.DEPEND_INDEX_DTYPE:
-                self._df.index = self._df.index.astype(define.DEPEND_INDEX_DTYPE)
+            self._df = self._set_dtypes(self._df)
 
         elif extension == "csv":
             table = csv.read_csv(
@@ -483,8 +475,7 @@ class Dependencies:
             values,
             columns=["file"] + list(define.DEPEND_FIELD_NAMES.values()),
         ).set_index("file")
-        df.index = df.index.astype(define.DEPEND_INDEX_DTYPE)
-
+        df = self._set_dtypes(df)
         self._df = pd.concat([self._df, df])
 
     def _add_meta(
@@ -583,6 +574,31 @@ class Dependencies:
         """
         self._df.at[file, "removed"] = 1
 
+    @staticmethod
+    def _set_dtypes(df: pd.DataFrame) -> pd.DataFrame:
+        r"""Set dependency table dtypes.
+
+        Args:
+            df: dataframe representing dependency table
+
+        Returns:
+            dataframe representing dependency table
+            with correct dtypes
+
+        """
+        # Check if we have an old cache,
+        # by checking the dtype of index,
+        # which changed to `object` in version 1.7.0
+        if df.index.dtype != define.DEPEND_INDEX_DTYPE:
+            df.index = df.index.astype(define.DEPEND_INDEX_DTYPE, copy=False)
+            columns = define.DEPEND_FIELD_NAMES.values()
+            dtypes = define.DEPEND_FIELD_DTYPES.values()
+            df = df.astype(
+                {name: dtype for name, dtype in zip(columns, dtypes)},
+                copy=False,
+            )
+        return df
+
     def _table_to_dataframe(self, table: pa.Table) -> pd.DataFrame:
         r"""Convert pyarrow table to pandas dataframe.
 
@@ -639,12 +655,7 @@ class Dependencies:
             values,
             columns=["file"] + list(define.DEPEND_FIELD_NAMES.values()),
         ).set_index("file")
-        df.index = df.index.astype(define.DEPEND_INDEX_DTYPE)
-        for name, dtype in zip(
-            define.DEPEND_FIELD_NAMES.values(),
-            define.DEPEND_FIELD_DTYPES.values(),
-        ):
-            df[name] = df[name].astype(dtype)
+        df = self._set_dtypes(df)
         self._df.loc[df.index] = df
 
     def _update_media_version(

--- a/audb/core/dependencies.py
+++ b/audb/core/dependencies.py
@@ -594,10 +594,8 @@ class Dependencies:
             df.index = df.index.astype(define.DEPEND_INDEX_DTYPE, copy=False)
             columns = define.DEPEND_FIELD_NAMES.values()
             dtypes = define.DEPEND_FIELD_DTYPES.values()
-            df = df.astype(
-                {name: dtype for name, dtype in zip(columns, dtypes)},
-                copy=False,
-            )
+            mapping = {column: dtype for column, dtype in zip(columns, dtypes)}
+            df = df.astype(mapping, copy=False)
         return df
 
     def _table_to_dataframe(self, table: pa.Table) -> pd.DataFrame:

--- a/tests/test_dependencies.py
+++ b/tests/test_dependencies.py
@@ -262,15 +262,66 @@ def test_load_save_backward_compatibility(tmpdir, deps):
     we need to make sure this is corrected
     when loading old cache files.
 
+    Old behaviour (audb<1.7):
+
+    archive          string[python]
+    bit_depth                 int32
+    channels                  int32
+    checksum         string[python]
+    duration                float64
+    format           string[python]
+    removed                   int32
+    sampling_rate             int32
+    type                      int32
+    version          string[python]
+
+    New behaviour (audb>=1.7):
+
+    archive          string[pyarrow]
+    bit_depth         int32[pyarrow]
+    channels          int32[pyarrow]
+    checksum         string[pyarrow]
+    duration         double[pyarrow]
+    format           string[pyarrow]
+    removed           int32[pyarrow]
+    sampling_rate     int32[pyarrow]
+    type              int32[pyarrow]
+    version          string[pyarrow]
+
     """
     deps_file = audeer.path(tmpdir, "deps.pkl")
+
+    deps_old = audb.Dependencies()
+    deps_old._df = deps._df.copy()
+
     # Change dtype of index from object to string
     # to mimic previous behavior
-    deps._df.index = deps._df.index.astype("string")
-    deps.save(deps_file)
+    deps_old._df.index = deps_old._df.index.astype("string")
+    # Change dtype of columns
+    # to mimic previous behavior
+    deps_old._df = deps_old._df.astype(
+        {
+            "archive": "string",
+            "bit_depth": "int32",
+            "channels": "int32",
+            "checksum": "string",
+            "duration": "float64",
+            "format": "string",
+            "removed": "int32",
+            "sampling_rate": "int32",
+            "type": "int32",
+            "version": "string",
+        }
+    )
+    deps_old.save(deps_file)
+
+    # Check that we get the correct dtypes,
+    # when loading from cache
     deps2 = audb.Dependencies()
     deps2.load(deps_file)
     assert deps2._df.index.dtype == audb.core.define.DEPEND_INDEX_DTYPE
+    pd.testing.assert_frame_equal(deps._df, deps2._df)
+    assert deps == deps2
 
 
 def test_load_save_errors(deps):

